Library/Homebrew/software_spec.rb

```diff
@@ -506,13 +506,15 @@ def compatible_locations?(tag: Utils::Bottles.tag)
 
     return true if RELOCATABLE_CELLARS.include?(cellar)
 
-    prefix = Pathname(cellar).parent.to_s
+    cellar = Pathname(cellar) # the required bottle cellar path may have symlinks on the user's machine
+    prefix = cellar.parent
 
-    cellar_relocatable = cellar.size >= HOMEBREW_CELLAR.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
-    prefix_relocatable = prefix.size >= HOMEBREW_PREFIX.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
+    cellar_relocatable = cellar.to_s.size >= HOMEBREW_CELLAR.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
+    prefix_relocatable = prefix.to_s.size >= HOMEBREW_PREFIX.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
 
-    compatible_cellar = cellar == HOMEBREW_CELLAR.to_s || cellar_relocatable
-    compatible_prefix = prefix == HOMEBREW_PREFIX.to_s || prefix_relocatable
+    # prefix/cellar may be symlinks but the physical paths are compatible
+    compatible_cellar = cellar.realpath.to_s == HOMEBREW_CELLAR.realpath.to_s || cellar_relocatable
```

Comment:
What will happen if this is a broken symlink?

Reply:
@XuehaiPan Or at least: preferable to adding the amount of code needed to check this.

Reply:
I just found we don't prefer symlinked cellar

https://github.com/Homebrew/brew/blob/f802815474fd84e4c81166ac9622bda3b24ac562/Library/Homebrew/diagnostic.rb#L524-L541

I would warn the user with more precise messages with this.

Reply:
> I would warn the user with more precise messages with this.

which user?

Reply:
For example, the users use symlinks for the Homebrew prefix. This would be widely common on Linux.

Reply:
If we're not getting support issues where that's the issue: I don't think we need to add to `brew doctor`.

Reply:
I add a warning to users where their cellar path contains symlinks.

Reply:
@XuehaiPan I don't understand what problem this is fixing?

Reply:
For Linux users, usually cloud service customers, the system admin sets a strict disk quota for the home directory. Usually, users have a small home directory and a large shared directory to store data, such as `/data/<username>` or `/share/<useraname>` depending on the IT services. They install Homebrew to the large disk the symlink it to a more convenient path.

If the bottle is built for the default cellar and not reallocatable, Homebrew pours bottles only when the following conditions are met:

- The cellar for the current brew installation is the default cellar (`/home/linuxbrew/.linuxbrew` on Linux).
- The cellar is a real directory that doesn't contain symlinks.

Homebrew reject pouring bottles for the Linux users I mentioned above. For example, the path `/home` or `/home/linuxbrew` could be a symlink due to the small disk quota.

Reply:
Sure, I guess I just don't understand what this PR is doing and how this behaviour would be discoverable.

> For Linux users, usually cloud service customers, the system admin sets a strict disk quota for the home directory. 

Can you point to some links of users complaining about this who don't have other workarounds?

Reply:
> I cannot find some links online. But I have received many complaints about "why Homebrew is building from the source all the time" from my colleagues offline (although it is not true that Homebrew also ships many formulae with reallocatable bottles). I have to explain that "this is how Homebrew works when no bottle is available/compatible".

👍🏻 fair.

I guess I'm wondering: how would these people be able to find this functionality if we add it? Does it need to be something we add in the installer?

> Allow using pre-built binaries always improves user experience, right? This PR allows pouring bottles that the user's cellar path has symlinks but logically it's equal to the default cellar.

How can they add these symlinks if they don't have the necessary access?

Reply:
> Can you point to some links of users complaining about this who don't have other workarounds?

That was my case. Workaround was to use a "bind mount" instead of a symbolic link but that might not be an option for everyone.